cmd/aws-s3-csi-controller/csicontroller/reconciler.go

```diff
@@ -119,20 +127,114 @@ func (r *Reconciler) reconcileMountpointPod(ctx context.Context, pod *corev1.Pod
 func (r *Reconciler) reconcileWorkloadPod(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
 	log := logf.FromContext(ctx).WithValues("pod", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name})
 
-	if pod.Spec.NodeName == "" {
-		log.V(debugLevel).Info("Pod is not scheduled to a node yet - ignoring")
+	if len(pod.Spec.Volumes) == 0 {
+		log.V(debugLevel).Info("Pod has no volumes - ignoring")
 		return reconcile.Result{}, nil
 	}
 
-	if len(pod.Spec.Volumes) == 0 {
-		log.V(debugLevel).Info("Pod has no volumes - ignoring")
+	scheduled := isPodScheduled(pod)
+	needsHeadroomForMountpointPod := r.isHeadroomForMountpointPodsFeatureEnabled() && mppod.ShouldReserveHeadroomForMountpointPod(pod)
+	hasHeadroomPods := r.isHeadroomForMountpointPodsFeatureEnabled() && mppod.WorkloadHasLabelPodForHeadroomPod(pod)
+	if !scheduled && !needsHeadroomForMountpointPod && !hasHeadroomPods {
+		log.V(debugLevel).Info("Pod is not scheduled to a node yet - ignoring")
 		return reconcile.Result{}, nil
 	}
 
-	var requeue bool
+	priorityClassKind := mppod.DefaultPriorityClass
+	if needsHeadroomForMountpointPod || hasHeadroomPods {
+		priorityClassKind = mppod.PreemptingPriorityClass
+	}
+
+	volumes, requeue, err := r.getWorkloadVolumes(ctx, pod)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
 	var errs []error
+	numHeadroomPods, numRemovedHeadroomPods := 0, 0
+
+	for i, vol := range volumes {
+		pv, pvc := vol.pv, vol.pvc
+
+		if scheduled {
+			log.V(debugLevel).Info("Found bound PV for PVC", "pvc", pvc.Name, "volumeName", pv.Name)
+
+			needsRequeue, err := r.spawnOrDeleteMountpointPodIfNeeded(ctx, pod, pvc, pv, priorityClassKind)
+			requeue = requeue || needsRequeue
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		} else if needsHeadroomForMountpointPod {
+			// Label the workload pod once for headroom pods
+			if i == 0 {
+				err = r.labelWorkloadPodForHeadroomPod(ctx, pod, log)
+				if err != nil {
+					errs = append(errs, err)
+					continue
+				}
+			}
+
+			err = r.createHeadroomPodForMountpointPod(ctx, pod, pv, log)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
 
-	for _, vol := range pod.Spec.Volumes {
+			numHeadroomPods += 1
+		}
+
+		if hasHeadroomPods && r.shouldDeleteHeadroomPodForTheWorkloadPod(pod) {
+			// Workload Pod past pending (i.e., either running or terminated),
+			// we no longer need the Headroom Pod
+			err := r.deleteHeadroomPodForMountpointPod(ctx, pod, pv, log)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			numRemovedHeadroomPods += 1
+		}
+	}
+
+	err = errors.Join(errs...)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if needsHeadroomForMountpointPod &&
+		len(volumes) == numHeadroomPods {
+		// Spawned all Headroom Pods needed, now ungate the Workload Pod,
+		// so it can get scheduled (alongside Headroom Pods)
+		err = r.ungateHeadroomSchedulingGateForWorkloadPod(ctx, pod, log)
```

Comment:
If headroom pod creation fails, then workload pod will be permanently in gated state, right? I.e. it won't be automatically moved into `Failed` state like before during `ContainerCreation` state. 

Should we add some troubleshooting guide on how to find and delete this stuck gated pods?